fv/netsets_test.go

```diff
@@ -0,0 +1,359 @@
+// +build fvtests
+
+// Copyright (c) 2017-2018 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fv_test
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"sync"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/projectcalico/felix/fv/utils"
+	"github.com/projectcalico/libcalico-go/lib/options"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/felix/fv/containers"
+	"github.com/projectcalico/felix/fv/workload"
+	api "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/net"
+)
+
+const (
+	numPolicies    = 20
+	numNetworkSets = 100
+)
+
+var _ = FContext("Network sets connectivity tests with initialized Felix and etcd datastore", func() {
+
+})
+
+var _ = Context("Network sets churn test with initialized Felix and etcd datastore", func() {
+
+	var (
+		etcd     *containers.Container
+		felix    *containers.Container
+		felixPID int
+		client   client.Interface
+	)
+
+	getFelixPIDs := func() []int {
+		return felix.GetPIDs("calico-felix")
+	}
+
+	updateFelixPID := func() {
+		// Get Felix's PID.  This retry loop ensures that we don't get tripped up if we see multiple
+		// PIDs, which can happen transiently when Felix restarts/forks off a subprocess.
+		start := time.Now()
+		for {
+			pids := getFelixPIDs()
+			if len(pids) == 1 {
+				felixPID = pids[0]
+				break
+			}
+			Expect(time.Since(start)).To(BeNumerically("<", time.Second))
+			time.Sleep(50 * time.Millisecond)
+		}
+	}
+
+	BeforeEach(func() {
+		felix, etcd, client = containers.StartSingleNodeEtcdTopology(containers.TopologyOptions{
+			FelixLogSeverity: "info",
+		})
+		updateFelixPID()
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			felix.Exec("iptables-save", "-c")
+		}
+		felix.Stop()
+
+		if CurrentGinkgoTestDescription().Failed {
+			etcd.Exec("etcdctl", "ls", "--recursive", "/")
```

Comment:
That won't give much for etcdv3...  I guess that's also a clue that this recurring logic should be utility-ized.

But perhaps those changes should be in another PR.  In that case best to leave this as is here, for ease of identifying it as one of the places that needs changing.

Reply:
(BTW, another Felix repo change that's needed in the same area is in calico-diags.  I'm planning to prepare that this week.)